[Frontend][Publisher][Redesign] Implement redesigned Data page

ivan-kishko opened the pull request:
## Description of the change

> Implement redesigned data page:
1. Redesign left menu
2. Redesign data view header 
3. Add table view
4. Add routing for metric-config page

## Type: Feature

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #135 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

ivan-kishko commented:
Just wanted to say, @mxosman thanks for a template (I tried different approach at first anyway but ended up using yours).

mxosman commented:
> Just wanted to say, @mxosman thanks for a template (I tried different approach at first anyway but ended up using yours).

Anytime, Ivan - I'm glad my messy template was useful in some way. Looking forward to diving into all of your changes - will do a more thorough review tomorrow morning.

hobuobi reviewed (changes requested):
Here's my review! Generally really impressed with this, most of these are small changes. @agronomic should also take a look at some of these. 
### Panel/Navigation (Left Side)
- Inactive headers should still be black; on hover, the plus sign should still appear. 
- Specific routes for each metric would be very helpful – I would love to be able to send a link to a specific metric, or even link to the Data Page for a metric from the Settings Page. 
- “Metric Configuration” link in the bottom-left corner should go to the Metric Configuration base page, not the page for the specific metric. 
- [idea] Animation for collapsing/opening metrics… it’s a little jarring right now. Still not the most important thing.
- [idea] Since the page isn’t supposed to scroll anyway, it would be cool to flip through metrics using the up/down arrow keys. 

### Data (Right Side)
- We need a better hover treatment for the “Switch to Data Table” and “Go to Metric Configuration” buttons. Even just reducing the opacity to 50% on hover would be nice. 
- We could actually expand the width this container – there’s a lot of empty space between the navigation and the data containers. Updated the designs accordingly. (@agronomic to also check this). 
- Header (everything above the line) should be sticky if the content is too tall. This shouldn't be necessary often, but for one of the metrics, there were so many rows on the data table that the page needed to scroll. 
#### For the Chart: 
- Long metric names or insights break the layout on normal screen widths. I don’t have a strong opinion on how to handle this right now (condense the Insights into a dropdown? supress them? move them?); curious if @agronomic or anyone else has ideas. 
- If there is no data: 
    - Suppress the "Go to Data Table" button.
    - Suppress the Insights (Most Recent, % Change, Avg. Total Value) 
#### For Data Table:
_Note: This was actually implemented to spec, but I want to suggest a few changes after playing with it. Proposed changes are represented in Figma [here](https://www.figma.com/file/dvX3yg7FMGJSWFaqHWrWgQ/Justice-Counts-Publisher?node-id=5470%3A15933&t=LQpnF6q4DjME5hec-0)._
- Bold the “Total” values (first row). 
- Add commas for thousands (e.g., 1000 -> 1,000)
- Add a line between breakdown sections (Total, Gender, Race/Ethnicity, etc) – the whitespace isn’t enough and it currently looks like a mess of numbers. 
- Add a hover state similar to what we did for the review page, except: 
    - Highlight the column instead of the row. 
    - Highlight the row only if you’re hovering over the heading (e.g., Total, Male, Other, etc)

terryttsai reviewed (commented):
This could be done as a follow-up clean-up PR, but I would like to see `updateActiveSystem`, `updateActiveMetricKey`, `getActiveSystemMetricKey`, `activeSystem`, and `activeMetricKey` removed from the store, since I don't think we need them anymore now that they're stored in the URL params!

terryttsai commented:
Everything looks really good otherwise—love how all these new components are organized!

mxosman commented on publisher/src/pages/Settings.tsx:
```diff
@@ -27,38 +27,29 @@ import {
-  "Metric Configuration",
-] as const;
-export type MenuOptions = typeof menuOptions[number];
+export const menuOptions = {
```
nit: I would rename this to something like `settingsMenuPaths` or something similar

ivan-kishko replied:
done [95eb95f](https://github.com/Recidiviz/justice-counts/pull/159/commits/95eb95fb47b8fa64cdcf250df12bcb5e79575ab4)

mxosman commented:
> Everything looks really good otherwise—love how all these new components are organized!

This is truly fantastic, Ivan! Love seeing how smoothly you implemented routing (took some notes myself)! Echoing Terry on having the params now be the single source of truth for the system and metric keys.

mxosman commented on publisher/src/components/Settings/types.ts:
```diff
@@ -0,0 +1,32 @@
+  const metric = params.get("metric") || undefined;
+
+  return { system, metric };
+};
```
small nit: up to you - I might move this to a `util`/`helper` file instead of the types

ivan-kishko replied:
done [95eb95f](https://github.com/Recidiviz/justice-counts/pull/159/commits/95eb95fb47b8fa64cdcf250df12bcb5e79575ab4)

ivan-kishko commented on common/components/DataViz/DatapointsTableView.tsx:
```diff
@@ -0,0 +1,242 @@
+  metricName?: string;
+  metricFrequency?: ReportFrequency;
+}> = ({ datapoints, useDataPageStyles, metricName, metricFrequency }) => {
+  if (!datapoints) return null;
```
change is in [7351fa7](https://github.com/Recidiviz/justice-counts/pull/159/commits/7351fa7ee104023e979195cd888c43124ce80398) - table view is unavailable if there is no data

ivan-kishko commented on publisher/src/components/MetricConfiguration/MetricBox.tsx:
```diff
@@ -44,14 +49,23 @@ export const MetricBox: React.FC<MetricBoxProps> = observer(
+    const settingsSearchParams = getSettingsSearchParams(searchParams);
+
+    const handleMetricBoxClick = () => {
+      updateActiveMetricKey(metricKey);
```
[573098a](https://github.com/Recidiviz/justice-counts/pull/159/commits/573098a0a3f87dfef3a20e720bf9b821a766f6e6)

ivan-kishko commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -70,7 +78,68 @@ export const MetricConfiguration: React.FC = observer(() => {
+    } else {
+      updateActiveMetricKey(undefined);
+    }
+    // updateActiveMetricKey(metricParam);
```
[573098a](https://github.com/Recidiviz/justice-counts/pull/159/commits/573098a0a3f87dfef3a20e720bf9b821a766f6e6)

ivan-kishko commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -70,7 +78,68 @@ export const MetricConfiguration: React.FC = observer(() => {
+          (system) => system === systemSearchParam
+        );
+      if (isUrlSystemParamInCurrentAgencySystems) {
+        updateActiveSystem(systemSearchParam);
```
[573098a](https://github.com/Recidiviz/justice-counts/pull/159/commits/573098a0a3f87dfef3a20e720bf9b821a766f6e6)

ivan-kishko commented:
Hello @terryttsai @mxosman. So I removed updateActiveSystem, updateActiveMetricKey, getActiveSystemMetricKey, activeSystem, activeMetricKey from store and its usages in components and made a custom hook to use settings search params inside corresponding components check out [573098a](https://github.com/Recidiviz/justice-counts/pull/159/commits/573098a0a3f87dfef3a20e720bf9b821a766f6e6) and [41a1c93](https://github.com/Recidiviz/justice-counts/pull/159/commits/41a1c93e214f8cffd6c6ce8e3c3275a72e35e79a). When #149 and #150 will be done eventually I will refactor MetricConfiguration component, added a TODO

ivan-kishko commented:
Hello again! @terryttsai @mxosman [b38177e](https://github.com/Recidiviz/justice-counts/pull/159/commits/b38177ed50ce5aea19e0c2867a3b23bc3d99c92a) - done here what we were talking about (activeMetricKey and activeSystem)

hobuobi commented:
This is fantastic!! Thank you so much for implementing those changes. 
One last thing: I found a bug (that exists in the current implementation as well) whereby someone can activate the "percentage" mode without any disaggregations active, and then have no way to switch back to "count" mode because a disaggregation has not been selected. Documented on [Loom](https://www.loom.com/share/ff0604ebf5dd4291b2304f3b67fa44ab).

If we could fix this before deploying that would be excellent! Other than that, I think we should be good to go.

ivan-kishko commented:
> This is fantastic!! Thank you so much for implementing those changes. One last thing: I found a bug (that exists in the current implementation as well) whereby someone can activate the "percentage" mode without any disaggregations active, and then have no way to switch back to "count" mode because a disaggregation has not been selected. Documented on [Loom](https://www.loom.com/share/ff0604ebf5dd4291b2304f3b67fa44ab).
> 
> If we could fix this before deploying that would be excellent! Other than that, I think we should be good to go.

Hello @hobuobi. About that bug - what exactly do you want the behavior to be - if no disaggregation selected automatically switch to "count view" or should control that sets the view type always be visible?

mxosman commented:
> Hello again! @terryttsai @mxosman [b38177e](https://github.com/Recidiviz/justice-counts/pull/159/commits/b38177ed50ce5aea19e0c2867a3b23bc3d99c92a) - done here what we were talking about (activeMetricKey and activeSystem)

Thank you for all the changes, Ivan! Code looks good to me - I'll let Humphrey give the final approval.

terryttsai reviewed (approved):
Thanks for making these changes! This looks great!

ivan-kishko commented:
> > This is fantastic!! Thank you so much for implementing those changes. One last thing: I found a bug (that exists in the current implementation as well) whereby someone can activate the "percentage" mode without any disaggregations active, and then have no way to switch back to "count" mode because a disaggregation has not been selected. Documented on [Loom](https://www.loom.com/share/ff0604ebf5dd4291b2304f3b67fa44ab).
> > If we could fix this before deploying that would be excellent! Other than that, I think we should be good to go.
> 
> Hello @hobuobi. About that bug - what exactly do you want the behavior to be - if no disaggregation selected automatically switch to "count view" or should control that sets the view type always be visible? I believe either is a quick fix just have to be sure what to do)

So I made for now that if you switch to "no disaggregation option" it switches back to count view. Let me know if you want to make it otherwise @hobuobi

lilidworkin commented:
@ivan-kishko this looks great to me! I think you have addressed all feedback really well. 

> So I made for now that if you switch to "no disaggregation option" it switches back to count view. Let me know if you want to make it otherwise

This is the right solution!

1. I have two small requests:
* Can we reduce the font size of the metric names in the left sidebar? It just feels so big to me:
<img width="423" alt="Screen Shot 2022-11-21 at 11 59 53 AM" src="https://user-images.githubusercontent.com/99899591/203115119-6b6ce138-d4e4-4575-b762-990c497903f9.png"> 
cc @hobuobi (but I'm sure he will be okay with this -- just pick a size that feels okay to you)


2. Can we add some minimum padding in the top bar between the MONTHLY/ANNUAL tag and the insights?
<img width="868" alt="Screen Shot 2022-11-21 at 12 00 40 PM" src="https://user-images.githubusercontent.com/99899591/203115278-9c8978ae-4685-4510-9f77-573273f71212.png">

hobuobi commented:
Agreed on that being the correct solution! And sorry for not specifying that :) 
Re: Lili's suggestions: 
1. I would actually push back on this one because I'd like the style of that sidebar to be the same as the Metric Configuration page. 
2. Agreed that this is still an issue, which arises because the horizontal space can't really expand beyond what's given. I would suggest the following: 
    - P0: White (opaque) background for the insights 
    - P0: Insights are always above (higher z-index) the metric name. 
    - P1: When hovering over the name, the insights disappear 
    - P2: The metric name automatically scrolls every few seconds so that you can see the whole thing, similar to what happens with a long song name on Spotify)

hobuobi commented:
Added a quick prototype to show what the ideal behavior for that might look like: https://www.figma.com/proto/dvX3yg7FMGJSWFaqHWrWgQ/Justice-Counts-Publisher?node-id=5784%3A13883&scaling=scale-down-width&page-id=1333%3A5802&starting-point-node-id=5784%3A13883
![longMetricName](https://user-images.githubusercontent.com/7134733/203198675-26b5def0-1130-4c81-b969-cf37634bf3c6.gif)

ivan-kishko commented:
> Added a quick prototype to show what the ideal behavior for that might look like: https://www.figma.com/proto/dvX3yg7FMGJSWFaqHWrWgQ/Justice-Counts-Publisher?node-id=5784%3A13883&scaling=scale-down-width&page-id=1333%3A5802&starting-point-node-id=5784%3A13883 ![longMetricName](https://user-images.githubusercontent.com/7134733/203198675-26b5def0-1130-4c81-b969-cf37634bf3c6.gif) [ ![longMetricName](https://user-images.githubusercontent.com/7134733/203198675-26b5def0-1130-4c81-b969-cf37634bf3c6.gif) ](https://user-images.githubusercontent.com/7134733/203198675-26b5def0-1130-4c81-b969-cf37634bf3c6.gif) [ ](https://user-images.githubusercontent.com/7134733/203198675-26b5def0-1130-4c81-b969-cf37634bf3c6.gif)

Hello @hobuobi! I think I got it but to be sure I would guess you want text to scroll ONLY if it is bigger than title container, and that being said you want me to calculate width of title container each time screen shrinks to identify should text scroll or not?

ivan-kishko commented:
Hello guys, @hobuobi @lilidworkin. So about that header problem, if I may, I would like to suggest in this commit my solution [69dc3bb]. I kind of combined initial thoughts and the recent request from you guys about it. What have I done: 
1. So I made fixed container for metric title of 700px (this should be enough for most cases). If title is bigger it will scroll like on gif Humphrey sent. Thing is that this container will be the same for table view as well. 
2. About insights - I calculate content width and make it disappear as soon as screen is smaller than left panel width title wrapper width and insight content width combined. it you hover over metric title you can see insights (done with title attribute)
3. About data view - I made min width of data view equal to title container width - this way title will behave always the same, and I think actually something like chart or table which contains a lot of data have to have some kind of a breakpoint it cannot be less than. User can just scroll to right if chart view overflowed.

And last thing. I did this because last idea had a few questions as well (I remember you said that we dont have to worry about small screens but still) like how to hover title if insights completely overflow them, should we always keep track of window width to adjust title wrapper and animation inside it. 

My solution is kind of in the middle as I said, please let me know what do you think about it.

lilidworkin reviewed (approved):
@ivan-kishko this looks fantastic, I'm really happy with your solution here! I think this is ready to merge. As discussed offline, I will ask @terryttsai  to resolve the merge conflicts for you and then merge it in. Thanks for all your hard work and patience on this!

terryttsai commented:
@hobuobi let me know if these changes are approved for you and then I can merge this PR!

hobuobi reviewed (approved):
All looking good from my POV! Thanks everyone for pulling through on this, especially @ivan-kishko for the detail-oriented work :)

lilidworkin commented:
Thank you so much @ivan-kishko , @terryttsai , and @hobuobi! What an awesome team effort all around!
